runtime/libc/wasi/include/wasi_serdes.h

```diff
@@ -0,0 +1,364 @@
+BASIC_TYPE_WASI(preopentype_t)
+#define WASI_SERDES_SIZE_riflags_t sizeof(__wasi_riflags_t)
+BASIC_TYPE_WASI(riflags_t)
+#define WASI_SERDES_SIZE_rights_t sizeof(__wasi_rights_t)
+BASIC_TYPE_WASI(rights_t)
+#define WASI_SERDES_SIZE_roflags_t sizeof(__wasi_roflags_t)
+BASIC_TYPE_WASI(roflags_t)
+#define WASI_SERDES_SIZE_sdflags_t sizeof(__wasi_sdflags_t)
+BASIC_TYPE_WASI(sdflags_t)
+#define WASI_SERDES_SIZE_siflags_t sizeof(__wasi_siflags_t)
+BASIC_TYPE_WASI(siflags_t)
+#define WASI_SERDES_SIZE_size_t sizeof(__wasi_size_t)
+BASIC_TYPE_WASI(size_t)
+#define WASI_SERDES_SIZE_inode_t sizeof(__wasi_inode_t)
+BASIC_TYPE_WASI(inode_t)
+#define WASI_SERDES_SIZE_signal_t sizeof(__wasi_signal_t)
+BASIC_TYPE_WASI(signal_t)
+#define WASI_SERDES_SIZE_subclockflags_t sizeof(__wasi_subclockflags_t)
+BASIC_TYPE_WASI(subclockflags_t)
+#define WASI_SERDES_SIZE_timestamp_t sizeof(__wasi_timestamp_t)
+BASIC_TYPE_WASI(timestamp_t)
+#define WASI_SERDES_SIZE_userdata_t sizeof(__wasi_userdata_t)
+BASIC_TYPE_WASI(userdata_t)
+#define WASI_SERDES_SIZE_whence_t sizeof(__wasi_whence_t)
+BASIC_TYPE_WASI(whence_t)
+
+#undef BASIC_TYPE_WASI
+#undef BASIC_TYPE
+
+/* WASI structure read/write functions. */
+
+#define STRUCT(name)                                                                     \
+    void wasi_serdes_write_##name(void* ptr, size_t offset, const __wasi_##name* value); \
+    void wasi_serdes_read_##name(const void* ptr, size_t offset, __wasi_##name* value);
+
+/* iovs currently only need to be read from WASM memory. */
+#define IOVS_STRUCT(name)                                                                            \
+    static inline __wasi_errno_t wasi_serdes_read_##name(const void* ptr, size_t end, size_t offset, \
+                                                         __wasi_##name* value);
+
+#define WASI_SERDES_SIZE_ciovec_t 8
+IOVS_STRUCT(ciovec_t)
+
+#define WASI_SERDES_SIZE_iovec_t 8
+IOVS_STRUCT(iovec_t)
+
+#define WASI_SERDES_SIZE_dirent_t 24
+STRUCT(dirent_t)
+
+#define WASI_SERDES_SIZE_fdstat_t 24
+STRUCT(fdstat_t)
+
+#define WASI_SERDES_SIZE_filestat_t 64
+STRUCT(filestat_t)
+
+#define WASI_SERDES_SIZE_prestat_t 8
+STRUCT(prestat_t)
+
+#define WASI_SERDES_SIZE_event_t 32
+STRUCT(event_t)
+
+#define WASI_SERDES_SIZE_subscription_t 48
+STRUCT(subscription_t)
+
+#undef STRUCT
+#undef IOVS_STRUCT
+
+static inline __wasi_errno_t
+wasi_serdes_readv_ciovec_t(const void* ptr, size_t end, size_t offset, __wasi_ciovec_t* iovs, __wasi_size_t iovs_len);
+
+static inline __wasi_errno_t
+wasi_serdes_readv_iovec_t(const void* ptr, size_t end, size_t offset, __wasi_iovec_t* iovs, __wasi_size_t iovs_len);
+
+/* Helper functions for memory bounds checking. */
+static inline int wasi_serdes_check_bounds(size_t offset, size_t end, size_t size);
+static inline int wasi_serdes_check_array_bounds(size_t offset, size_t end, size_t size, size_t count);
+
+static inline int wasi_serdes_check_bounds(size_t offset, size_t end, size_t size) {
```

Comment:
Did the formatter put all of this on the same line? Does awsm not use the "type on line alone" style? Sorry I completely forget.

Reply:
I think that I'm coming to understand that you likely got a lot of this logic from elsewhere? If so, make sure to acknowledge that in the header (and maintain license).

Reply:
The aWsm clang-format is configured to place everything on one line to match Gregor's "Rusty" preferences. The SLEdge config uses "type on line alone."